lib/nrf_modem_lib/nrf91_sockets.c

```diff
@@ -810,12 +810,22 @@ static ssize_t nrf91_socket_offload_recvfrom(void *obj, void *buf, size_t len,
 		}
 
 		if (cliaddr->sa_family == NRF_AF_INET) {
-			nrf_to_z_ipv4(from, (struct nrf_sockaddr_in *)cliaddr);
+			struct sockaddr_in tmp_from;
+
+			nrf_to_z_ipv4((struct sockaddr *)&tmp_from,
```

Comment:
This function copies `cliaddr` into `tmp_from`, however, we don't check if any bytes have been written to `cliaddr` by `recvfrom` (the value of `sock_len`). If the sender address is not available (as is the case for DTLS), then checking for `cliaddr->sa_family` is wrong because the contents of `cliaddr` are unspecified.

It is possible that `cliaddr->sa_family` is `NRF_AF_INET` (since it's unspecified) and the rest is garbage, which is copied into `tmp_from` by `nrf_to_z_ipv4`. The `memcpy` will then copy up to `MIN(*fromlen, sizeof(struct sockaddr_in))` bytes of that unspecified contents, into `from` (which is returned to the user).

And here's the second issue, we don't check that `fromlen` is `NULL`. This is coming from the user, and it is not passed to `recvfrom` (which would emit -EFAULT in that case). Instead, we dereference it. That's a problem because if the user passes `NULL` the function will crash; if it is not `NULL`, we will copy that many bytes of garbage into the output param `from`.

Reply:
> It is possible that cliaddr->sa_family is NRF_AF_INET (since it's unspecified) and the rest is garbage, which is copied into tmp_from by nrf_to_z_ipv4. The memcpy will then copy up to MIN(*fromlen, sizeof(struct sockaddr_in)) bytes of that unspecified contents, into from (which is returned to the user).

As the cliaddr is zero'ed out, this should not happen, but I added this check.

> And here's the second issue, we don't check whether fromlen is NULL. This is coming from the user, and it is not passed to recvfrom (which would return an error in that case). Instead, we dereference it. That's a problem because if the user passes NULL the function will crash; if it is not NULL, we will copy that many bytes of garbage into the output param from.

Added a check for this.